In [63]:
import numpy as np, pandas as pd, matplotlib.pyplot as plt, mlflow
from sklearn.datasets import fetch_california_housing
from sklearn.base import clone
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.linear_model import LinearRegression, LogisticRegression, Ridge, Lasso
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline


In [ ]:
data = fetch_california_housing(as_frame=True)
X = data.data
y = data.target
pipe = make_pipeline(StandardScaler(), LinearRegression())

scores = cross_val_score(pipe, X, y, cv=5, scoring='r2')
print(scores)
print(scores.mean())
print(scores.std())
f"{scores.mean():.4f} ± {scores.std():.4f}"

[0.54866323 0.46820691 0.55078434 0.53698703 0.66051406]
0.5530311140279563
0.061691601409532774


'0.5530 ± 0.0617'

From single split the r2score is : 0.5757, and from cv, mean r2score is 0.5530 and std is 0.0616 which >0.05, as the std is more which means that model is unstable and we can not relay on it

In [72]:
pipe = make_pipeline(StandardScaler(), Ridge(1.0))

scores = cross_val_score(pipe, X, y, cv=5, scoring='r2' )
print(scores)
print(scores.mean())
print(scores.std())
f"{scores.mean():.4f} ± {scores.std():.4f}"

[0.54874936 0.46818182 0.55080227 0.53691869 0.66051907]
0.55303424082049
0.0617024714315415


'0.5530 ± 0.0617'

For Ridge regression, From the single split train test dataset, the r2score is 0.5758 and from cv r2score is 0.5530 and std is 0.0617 which is more (>0.05) and the model is still less stable.

In [73]:
pipe = make_pipeline(StandardScaler(), Lasso(0.1))
scores = cross_val_score(pipe, X, y, cv=5, scoring='r2' )
print(scores)
print(scores.mean())
print(scores.std())
print(f"{scores.mean():.4f} ± {scores.std():.4f}")

[0.45509221 0.40234689 0.47091764 0.30468546 0.52259382]
0.43112720517469005
0.0739453362418414
0.4311 ± 0.0739


For Lasso regression, from the single split train-test dataset, the value is 0.4813 and from cv is 0.4311 and std is 0.0739 which is good as it less than 0.05 and with ridge regression we can say that the model stable when compared with ridge model